### PR TITLE
use ShouldGrowVolumesByDcAndRack

### PR DESCRIPTION
**Pull request by kmlebedev**
# What problem are we solving?

https://github.com/seaweedfs/seaweedfs/issues/6278

# How are we solving the problem?

We check the DC and the Rack

# How is the PR tested?

# Checks
- [ ] I have added unit tests if possible.
- [ ] I will add related wiki document changes and link to this PR after merging.

### CODE REVIEW

File: weed/topology/topology.go [hunk 1]
```diff
@@ -369,6 +369,19 @@ func (t *Topology) ListDataCenters() (dcs []string) {
 	return dcs
 }
 
+func (t *Topology) ListDCAndRacks() (dcs map[string][]string) {
```
Comment (chrislusf) [inline weed/topology/topology.go]: Use types instead of strings

File: weed/topology/volume_layout.go [hunk 2]
```diff
@@ -399,6 +386,14 @@ func (vl *VolumeLayout) GetWritableVolumeCount() (active, crowded int) {
 	return len(vl.writables), len(vl.crowded)
 }
 
+func (vl *VolumeLayout) GetWritableVolumes() (writables []needle.VolumeId) {
```
Comment (chrislusf) [inline weed/topology/volume_layout.go]:
```suggestion
func (vl *VolumeLayout) CloneWritableVolumes() (writables []needle.VolumeId) {
```

Comment (kmlebedev) [reply weed/topology/volume_layout.go]: ok

File: weed/topology/topology.go (hunk 1 above)
Comment (kmlebedev) [reply weed/topology/topology.go]: > Use types instead of strings

done

**Comment by chrislusf** [comment]
How about simplify this by adding a structure to each volume layout, tracking sent volume grow requests that are not reported back by the volume servers?

**Comment by kmlebedev** [comment]
> How about simplify this by adding a structure to each volume layout, tracking sent volume grow requests that are not reported back by the volume servers?

@chrislusf 
I didn't quite understand how this would help.
But now, there is a problem in the synchronous creation of volumes where there are none, precisely because of the lack of a structure for each layer with a list of volumes available for writing. Because it's difficult to tie it together.

Maybe later I'll make another approach to this

**Pull request merged**